### ISSUE TITLE
Rule for requiring class final

### ISSUE THREAD

**Issue by Igor-Palaguta** [realm/SwiftLint#228]
It will be cool if you can create rule for requiring for class ```final``` if it is not used as parent. As it affects performance

Also (lower priority) ```private``` can be required if code used just inside module

**Comment by jpsim** [comment]
There's no way for us to know if a `public` class is designed to be subclassed (even though it's not subclassed in the module in which it's defined). Take [`RealmSwift.Object`](https://realm.io/docs/swift/latest/api/Classes/Object.html) for example.

However, if a class is `internal` or `private` and not subclassed anywhere within the module, then that could trigger a violation.

> Also (lower priority) `private` can be required if code used just inside module

See my first point, it's not possible to know if consumers of a framework will use `public` code.

**Comment by Igor-Palaguta** [comment]
For public you are right, but for internal to display warning is very cool. If user does not want this warn they can always disable it.

**Comment by Igor-Palaguta** [comment]
Btw if module is application (not framework), not sure public or internal plays role

**Comment by jpsim** [comment]
If the module is an application, then using `public` is semantically incorrect IMO.

**Comment by Igor-Palaguta** [comment]
I mean that swiflint can easily suggest final or unknown in most part of cases. If it is hard to do, or you think not so important ignore this. But every time I search manually for such places and add final manually. You can simplify this and make our app with minimal effort faster

**Comment by Igor-Palaguta** [comment]
Probably I ask a lot as swiftlint search just for stylistic issues. Ignore this request. swiftlint is not static analyzer

**Comment by masters3d** [comment]
Now that we have sealed public classes by default, I don't see the performance benefit for this rule in swift3
https://github.com/apple/swift-evolution/blob/master/proposals/0117-non-public-subclassable-by-default.md

**Comment by jpsim** [comment]
Agreed. Closing.
